File: python/leetcode/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        open_bracket = ['(', '[', '{']
        closed_bracket = [')', ']', '}']
        
        if len(s) <= 1:
            return False
        
        stack = []
        
        for i in range(len(s)):
            if s[i] in open_bracket:
                stack.append(s[i])
            elif s[i] in closed_bracket:
                if len(stack) == 0:
                    
                    return False
                
                top = stack.pop()
                if s[i] == ')' and top != '(':
                    return False
                if s[i] == ']' and top != '[':
                    return False
                if s[i] == '}' and top != '{':
                    return False
                    
        return len(stack) == 0
                    
        

File: c#/leetcode/20-valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        Stack<char> stack = new Stack<char>();
        
        for (int i=0; i < s.Length; i++) {
            if(s[i] == '(' || s[i] == '[' || s[i] == '{') {
                stack.Push(s[i]);
            }
            else {
                if (stack.Count == 0) {
                    return false;
                }
                char last = stack.Pop();
                if (s[i] == ')' && last != '(') {
                    return false;
                }
                if (s[i] == ']' && last != '[') {
                    return false;
                }
                if(s[i] == '}' && last != '{') {
                    return false;
                }
            }
        }
        return stack.Count == 0;      
    }
}